metals/src/main/scala/scala/meta/internal/bsp/BspConnector.scala

```diff
@@ -137,9 +137,26 @@ class BspConnector(
         case ResolvedBspOne(details) =>
           tables.buildServers.chooseServer(details.getName())
           optSetBuildTool(details.getName())
-          bspServers
-            .newServer(projectRoot, bspTraceRoot, details, bspStatusOpt)
-            .map(Some(_))
+          buildTool match {
+            case Some(bsp: BuildServerProvider)
+                if bsp.shouldRegenerateBspJson(details.getVersion()) =>
+              bsp
+                .generateBspConfig(
+                  workspace,
+                  args => bspConfigGenerator.runUnconditionally(bsp, args),
+                  statusBar,
+                )
+                .map(status => handleGenerationStatus(bsp, status))
+                .flatMap { _ =>
+                  connect(projectRoot, bspTraceRoot, addLivenessMonitor)
```

Comment:
Not sure if that is an issue but it looks like if things go wrong this can loop forever.